src/phoenix/server/api/routers/trace_handler.py

```diff
@@ -54,7 +60,122 @@ async def post(self, request: Request) -> Response:
         for resource_spans in req.resource_spans:
             project_name = get_project_name(resource_spans.resource.attributes)
             for scope_span in resource_spans.scope_spans:
-                for span in scope_span.spans:
-                    self.traces.put(decode(span), project_name=project_name)
+                for otlp_span in scope_span.spans:
+                    span = decode(otlp_span)
+                    async with self.db() as session:
+                        await _insert_span(session, span, project_name)
+                    self.traces.put(span, project_name=project_name)
                     await asyncio.sleep(0)
         return Response()
+
+
+async def _insert_span(session: AsyncSession, span: Span, project_name: str) -> None:
+    if not (
+        project_rowid := await session.scalar(
+            text("SELECT rowid FROM projects WHERE name = :name;"),
+            {"name": project_name},
+        )
+    ):
+        project_rowid = await session.scalar(
+            text("INSERT INTO projects(name) VALUES(:name) RETURNING rowid;"),
+            {"name": project_name},
+        )
+    if (
+        trace_rowid := await session.scalar(
+            text(
+                """
+                INSERT INTO traces(trace_id, project_rowid, session_id, start_time, end_time)
+                VALUES(:trace_id, :project_rowid, :session_id, :start_time, :end_time)
+                ON CONFLICT DO UPDATE SET
+                start_time = CASE WHEN excluded.start_time < start_time THEN excluded.start_time ELSE start_time END,
+                end_time = CASE WHEN end_time < excluded.end_time THEN excluded.end_time ELSE end_time END
+                WHERE excluded.start_time < start_time OR end_time < excluded.end_time
```

Comment:
I'm maybe starting to not fully groc the need to manage tracing timing info like this. Why can't the spans be the source of truth for these? Need a refresher - sorry.

Reply:
no, it's not required. it's just an optimization